l2-contracts/contracts/ConsensusRegistry.sol

```diff
@@ -0,0 +1,478 @@
+            }
+        }
+
+        nodeOwners.push(_nodeOwner);
+        nodes[_nodeOwner] = Node({
+            attesterLatest: AttesterAttr({
+            active: true,
+            weight: _attesterWeight,
+            pubKey: _attesterPubKey
+        }),
+            attesterSnapshot: AttesterAttr({
+            active: false,
+            weight: 0,
+            pubKey: Secp256k1PublicKey({a: bytes1(0), b: bytes16(0)})
+        }),
+            attesterLastUpdateCommit: attestersCommit,
+            validatorLatest: ValidatorAttr({
+            active: true,
+            weight: _validatorWeight,
+            pubKey: _validatorPubKey,
+            pop: _validatorPoP
+        }),
+            validatorSnapshot: ValidatorAttr({
+            active: false,
+            weight: 0,
+            pubKey: BLS12_381PublicKey({a: bytes32(0), b: bytes32(0), c: bytes32(0)}),
+            pop: BLS12_381Signature({a: bytes32(0), b: bytes16(0)})
+        }),
+            validatorLastUpdateCommit: validatorsCommit
+        });
+
+        emit NodeAdded(
+            _nodeOwner,
+            _validatorWeight,
+            _validatorPubKey,
+            _validatorPoP,
+            _attesterWeight,
+            _attesterPubKey
+        );
+    }
+
+    /// @notice Deactivates a node, preventing it from participating in committees.
+    /// @dev Only callable by the contract owner or the node owner.
+    /// @dev Verifies that the node owner exists in the registry.
+    /// @param _nodeOwner The address of the node's owner to be inactivated.
+    function deactivate(address _nodeOwner) external onlyOwnerOrNodeOwner(_nodeOwner) {
+        verifyNodeOwnerExists(_nodeOwner);
+
+        Node storage node = nodes[_nodeOwner];
+        ensureAttesterSnapshot(node);
+        node.attesterLatest.active = false;
+        ensureValidatorSnapshot(node);
+        node.validatorLatest.active = false;
+
+        emit NodeDeactivated(_nodeOwner);
+    }
+
+    /// @notice Activates a previously inactive node, allowing it to participate in committees.
+    /// @dev Only callable by the contract owner or the node owner.
+    /// @dev Verifies that the node owner exists in the registry.
+    /// @param _nodeOwner The address of the node's owner to be activated.
+    function activate(address _nodeOwner) external onlyOwnerOrNodeOwner(_nodeOwner) {
+        verifyNodeOwnerExists(_nodeOwner);
+
+        Node storage node = nodes[_nodeOwner];
+        ensureAttesterSnapshot(node);
+        node.attesterLatest.active = true;
+        ensureValidatorSnapshot(node);
+        node.validatorLatest.active = true;
+
+        emit NodeActivated(_nodeOwner);
+    }
+
+    /// @notice Removes a node from the registry.
+    /// @dev Only callable by the contract owner.
+    /// @dev Verifies that the node owner exists in the registry.
+    /// @param _nodeOwner The address of the node's owner to be removed.
+    function remove(address _nodeOwner) external onlyOwner {
```

Comment:
What if the owner call this function twice? This will add node removal twice to the `pendingRemoval` list, which is not great. Moreover `commitValidators` and `commitAttesters` will stop working `finalizeNodeRemoval`. Here is why:
- two identical pending `_nodeOwner`s are on the `pendingRemovals` array.
- `commitAttesters` is called, now `pendingAttestersCommit` is `false` for both elements
- `pendingValidatorsCommit` is called, now `pendingValidatorsCommit` is `false` for both elements
- on the `pendingValidatorsCommit` the function tries to do `finalizeNodeRemoval`. For the first element it works, for the second it doesn't because element was deleted already.

Btw, this is a clear example of why `traversing` array is bad in one function. If there is an revert because of one element the whole function stops working.

Reply:
I applied your proposed solution instead (here: https://github.com/matter-labs/era-contracts/pull/555/commits/ec3699373456fb30747a13c34af702dcccc5f244), so that the finalization of the removal can be triggered by any function. 
I originally argued against it thinking it will lack incentive compatibility for bearing the gas cost of that removal. But I agree that making the current solution secure is probably not worthwhile, gas cost-wise.